### PR TITLE
OTF encoder

### PR DESCRIPTION
**Pull request by camertron**
_**This pull request is part of a larger effort to bring OTF support to TTFunk. See https://github.com/prawnpdf/ttfunk/issues/53 for details.**_

This PR adds the `OTFEncoder` class. There's really no difference between .otf and .ttf fonts except that CFF-based fonts (often given the .otf extension) have a different optimal table order.I

Things to note in this PR:
1. I went through and changed string encoding in a couple of places to make sure there is uniform encoding throughout the CFF code. Since we're packing binary data, the strings emitted should always have ASCII-8BIT encoding.
2. Supplemental CFF encoding support has been added.
3. Added support for the gasp table.
4. Fixed a number of issues I found after parsing around 10,000 real fonts.

@pointlessone unfortunately this won't be the last PR. I have a couple more in the works that I will add to #53. I also just re-tested subsetting Noto, and while all the ASCII characters render in Photoshop, Keynote, etc, none of the Japanese characters work (I just see blank spaces). So unfortunately something's wrong. I will need to identify and fix the issue before we can consider the OTF work done.

### CODE REVIEW

File: lib/ttfunk/table/cff.rb [hunk 1]
```diff
@@ -31,18 +31,18 @@ def tag
         TAG
       end
 
-      def encode(mapping)
+      def encode(new2_old, old2_new)
```
Comment (pointlessone) [inline lib/ttfunk/table/cff.rb]: I don't quite get this change. Both mapping appear to always be derivable from another. Why do we need to pass around both?

File: lib/ttfunk/table/cff/encoding.rb [hunk 2]
```diff
@@ -89,8 +92,30 @@ def encode(mapping)
...
         end
 
+        def supplemental?
+          # high-order bit set to 1 indicates supplemental encoding
+          @format >> 7 == 1
+        end
+
         private
 
+        def encode_supplemental(_new2_old, old2_new)
```
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/table/cff/encoding.rb]: I think I named these this way because of existing code, should we clean it up everywhere?

File: lib/ttfunk/table/cff.rb (hunk 1 above)
Comment (camertron) [reply lib/ttfunk/table/cff.rb]: Yes they are derivable from each other, but passing them around is much more efficient than requiring each `encode` method to re-create them.

File: lib/ttfunk/table/maxp.rb [hunk 3]
```diff
@@ -50,6 +52,10 @@ def parse!
             @max_storage, @max_function_defs, @max_instruction_defs,
             @max_stack_elements, @max_size_of_instructions,
             @max_component_elements, @max_component_depth = read(28, 'Nn*')
+
+          # a number of fonts omit these last two bytes for some reason,
```
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/table/maxp.rb]: I believe so, yes. I saw the `nil` issue in fonts where the maxp table was the last table in the font. We do have the length here so it would not be too difficult to protect against reading past the end of the table.

File: lib/ttfunk/table/cff/dict.rb [hunk 4]
```diff
@@ -177,8 +177,8 @@ def decode_operand(b_zero)
         end
 
         def decode_sci
-          significand = []
```
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/table/cff/dict.rb]: In this case I want a binary string, but I also want it to be mutable. Not sure I understand what your concern is here... mutating strings is not inherently bad is it?

File: lib/ttfunk/encoded_string.rb [hunk 5]
```diff
@@ -51,15 +51,15 @@ def string
           "#{placeholders.size} unresolved placeholder(s)"
       end
 
-      io.string
+      io.string.force_encoding(Encoding::ASCII_8BIT)
```
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/encoded_string.rb]: Ok that makes sense. As I recall `StringIO` doesn't respect the `b` option if you pass it in as the `mode`, which also could have been a solution here.

File: lib/ttfunk/table/cff/encoding.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (pointlessone) [reply lib/ttfunk/table/cff/encoding.rb]: Since we're changing this code substantially we might as well clean it up a bit.

File: lib/ttfunk/table/maxp.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (pointlessone) [reply lib/ttfunk/table/maxp.rb]: Let's do that.

File: lib/ttfunk/table/cff.rb (hunk 1 above)
Comment (pointlessone) [reply lib/ttfunk/table/cff.rb]: I see. Please document this somewhere. Either a comment in source or commit message.

File: lib/ttfunk/table/cff/dict.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (pointlessone) [reply lib/ttfunk/table/cff/dict.rb]: It is not inherently bad to mutate a string. However, I'm trying to avoid it in the codebase for consistency.

**Comment by camertron** [comment]
Ok @pointlessone I've addressed all your comments and fixed subsetting too. Code can subset Noto and resulting font can be used by word processors, etc.

File: lib/ttfunk/table/cff.rb (hunk 1 above)
Comment (camertron) [reply lib/ttfunk/table/cff.rb]: Done

File: lib/ttfunk/table/maxp.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/table/maxp.rb]: Done (and found a bug as a result!)

File: lib/ttfunk/table/cff/encoding.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (camertron) [reply lib/ttfunk/table/cff/encoding.rb]: Done (renamed to `old_to_new` and `new_to_old`).
